src/core/thread/mle_router.cpp

```diff
@@ -3494,5 +3546,128 @@ ThreadError MleRouter::AppendPendingDataset(Message &aMessage)
     return error;
 }
 
+bool MleRouter::HasMinDowngradeNeighborRouters(void)
+{
+    return GetMinDowngradeNeighborRouters() >= kMinDowngradeNeighbors;
+}
+
+bool MleRouter::HasOneNeighborwithComparableConnectivity(const RouteTlv &aRoute, uint8_t aRouterId)
+{
+    uint8_t localLqi = 0;
+    uint8_t peerLqi = 0;
+    uint8_t routerCount = 0;
+
+    // process local neighbor routers
+    for (uint8_t i = 0; i < kMaxRouterId; i++)
+    {
+        if (i == mRouterId)
+        {
+            routerCount++;
+            continue;
+        }
+
+        // check if neighbor is valid
+        if (mRouters[i].mState == Neighbor::kStateValid)
+        {
+            // if peer is just neighbor
+            if (i == aRouterId)
+            {
+                routerCount++;
+                continue;
+            }
+
+            localLqi = mRouters[i].mLinkQualityOut;
+
+            if (localLqi > mRouters[i].mLinkInfo.GetLinkQuality(mMac.GetNoiseFloor()))
+            {
+                localLqi = mRouters[i].mLinkInfo.GetLinkQuality(mMac.GetNoiseFloor());
+            }
+
+            if (localLqi == 2 || localLqi == 3)
+            {
+                // check if this neighbor router is in peer Route64 TLV
+                if (aRoute.IsRouterIdSet(i) == false)
+                {
+                    return false;
+                }
+
+                // get the peer's two-way lqi to this router
+                peerLqi = aRoute.GetLinkQualityOut(routerCount);
+
+                if (peerLqi > aRoute.GetLinkQualityIn(routerCount))
+                {
+                    peerLqi = aRoute.GetLinkQualityIn(routerCount);
+                }
+
+                // compare local lqi to this router with peer's
+                if (peerLqi >= localLqi)
+                {
+                    routerCount++;
+                    continue;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            routerCount++;
+        }
+    }
+
+    return true;
+}
+
+bool MleRouter::HasSmallNumberOfChildren(void)
+{
+    Child *children;
+    uint8_t maxChildCount = 0;
+    uint8_t numChildren = 0;
+    uint8_t routerCount = GetActiveRouterCount();
+
+    VerifyOrExit(routerCount > mRouterDowngradeThreshold, ;);
+
+    children = GetChildren(&maxChildCount);
+
+    for (uint8_t i = 0; i < maxChildCount; i++)
+    {
+        if (children[i].mState == Neighbor::kStateValid)
+        {
+            numChildren++;
+        }
+    }
+
+    return numChildren < (routerCount - mRouterDowngradeThreshold) * 3;
+
+exit:
+    return false;
+}
+
+uint8_t MleRouter::GetMinDowngradeNeighborRouters(void)
+{
+    uint8_t lqi;
+    uint8_t routerCount = 0;
+
+    for (uint8_t i = 0; i < kMaxRouterId; i++)
+    {
+        if (mRouters[i].mState == Neighbor::kStateValid)
```

Comment:
style: checking `mState != kStateValid` and using `continue` here would minimize indenting a large block below.